cpp/src/parquet/decoder.cc

```diff
@@ -654,43 +701,41 @@ class PlainByteArrayDecoder : public PlainDecoder<ByteArrayType>,
                           int64_t valid_bits_offset,
                           typename EncodingTraits<ByteArrayType>::Accumulator* out,
                           int* out_values_decoded) {
-    ArrowBinaryHelper<ByteArrayType> helper(out, num_values);
-    int values_decoded = 0;
-
-    RETURN_NOT_OK(helper.Prepare(len_));
-
-    int i = 0;
-    RETURN_NOT_OK(VisitNullBitmapInline(
-        valid_bits, valid_bits_offset, num_values, null_count,
-        [&]() {
-          if (ARROW_PREDICT_FALSE(len_ < 4)) {
-            ParquetException::EofException();
-          }
-          auto value_len = SafeLoadAs<int32_t>(data_);
-          if (ARROW_PREDICT_FALSE(value_len < 0 || value_len > INT32_MAX - 4)) {
-            return Status::Invalid("Invalid or corrupted value_len '", value_len, "'");
-          }
-          auto increment = value_len + 4;
-          if (ARROW_PREDICT_FALSE(len_ < increment)) {
-            ParquetException::EofException();
-          }
-          RETURN_NOT_OK(helper.PrepareNextInput(value_len, len_));
-          helper.UnsafeAppend(data_ + 4, value_len);
-          data_ += increment;
-          len_ -= increment;
-          ++values_decoded;
-          ++i;
-          return Status::OK();
-        },
-        [&]() {
-          helper.UnsafeAppendNull();
-          ++i;
-          return Status::OK();
-        }));
+    auto visit_binary_helper = [&](auto* helper) {
+      int values_decoded = 0;
+
+      RETURN_NOT_OK(VisitNullBitmapInline(
+          valid_bits, valid_bits_offset, num_values, null_count,
+          [&]() {
+            if (ARROW_PREDICT_FALSE(len_ < 4)) {
+              return Status::Invalid(
+                  "Invalid or truncated PLAIN-encoded BYTE_ARRAY data");
+            }
+            auto value_len = SafeLoadAs<int32_t>(data_);
+            if (ARROW_PREDICT_FALSE(value_len < 0 || value_len > len_ - 4)) {
+              return Status::Invalid(
+                  "Invalid or truncated PLAIN-encoded BYTE_ARRAY data");
+            }
+            RETURN_NOT_OK(helper->AppendValue(data_ + 4, value_len,
+                                              /*estimated_remaining_data_length=*/len_));
```

Comment:
This is so confusing to me. `AppendValue` continously with `len_` (nearing the "whole" page size) would be heavy?

Reply:
Yes, the API could look weird. You have to take a look at the `AppendValue` definition so that it makes sense: the `len_` is only used when a new chunk is started, to reserve the data on the new builder.

Reply:
Aha, I got this, AppendValue with length means the chunk size, this is really tricky